File: 173.Binary-Search-Tree-Iterator.go
// https://leetcode.com/problems/binary-search-tree-iterator/
//
// algorithms
// Medium (45.79%)
// Total Accepted:    172,665
// Total Submissions: 377,082
// beats 100.0% of golang submissions

package leetcode

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type BSTIterator struct {
	valArr []int
	cntIdx int
}

func Constructor(root *TreeNode) BSTIterator {
	var valArr []int
	resursive(root, &valArr)
	return BSTIterator{
		valArr: valArr,
		cntIdx: 0,
	}
}

func resursive(root *TreeNode, valArr *[]int) {
	if root == nil {
		return
	}
	resursive(root.Left, valArr)
	*valArr = append(*valArr, root.Val)
	resursive(root.Right, valArr)
}

/** @return the next smallest number */
func (this *BSTIterator) Next() int {
	res := this.valArr[this.cntIdx]
	this.cntIdx++

	return res
}

/** @return whether we have a next smallest number */
func (this *BSTIterator) HasNext() bool {
	return this.cntIdx < len(this.valArr)
}

/**
 * Your BSTIterator object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Next();
 * param_2 := obj.HasNext();
 */

File: 173.Binary-Search-Tree-Iterator.py
# https://leetcode.com/problems/binary-search-tree-iterator/
#
# algorithms
# Medium (49.5%)
# Total Accepted:    215,089
# Total Submissions: 434,545


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

from collections import deque


class BSTIterator(object):

    def __init__(self, root):
        """
        :type root: TreeNode
        """
        self.stack = deque()
        self.push(root)

    def next(self):
        """
        @return the next smallest number
        :rtype: int
        """
        top = self.stack.pop()
        self.push(top.right)

        return top.val

    def hasNext(self):
        """
        @return whether we have a next smallest number
        :rtype: bool
        """
        return bool(self.stack)

    def push(self, root):
        while root:
            self.stack.append(root)
            root = root.left


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()
